File: javascript/grammar.js
var Grammar = (function() {
  var PRONOUNS = {
    male: {
      himself: "himself",
      his: "his",
      him: "him",
      he: "he"
    },
    female: {
      himself: "herself",
      his: "her",
      him: "her",
      he: "she"      
    },
    neuter: {
      himself: "itself",
      his: "its",
      him: "it",
      he: "it"      
    },
    you: {
      himself: "yourself",
      his: "your",
      him: "you",
      he: "you"      
    }
  };

  var utils = {
    extend: function(a, b) {
      for (var name in b)
        a[name] = b[name];
      return a;
    },
    makePronounDict: function(gender, prefix) {
      if (!prefix)
        prefix = "";
      if (!(gender in PRONOUNS))
        throw new Error("invalid gender: " + gender);
      var dict = {};
      for (var name in PRONOUNS[gender])
        dict[prefix + name] = PRONOUNS[gender][name];
      return dict;
    },
    applyDict: function(string, dict) {
      return string.replace(/%\(([a-zA-Z0-9_]+)\)s/g, function(str, term) {
        return dict[term].toString();
      });
    },
    parseString: function(string, isSquarePlural, isAnglePlural) {
      var SQUARES = /\[(.*?)(\|(.+?))?\]/g,
          ANGLES  = /\<(.*?)(\|(.+?))?\>/g;

      function replaceBrackets(string, regexp, isPlural) {
        return string.replace(regexp, function(str, singular, plural) {
          if (isPlural)
            return plural ? plural.slice(1) : "";
          return singular;
        });
      }

      string = replaceBrackets(string, SQUARES, isSquarePlural);
      return replaceBrackets(string, ANGLES, isAnglePlural);
    }
  };

  return {
    utils: utils,
    View: function View(viewer) {
      var self = {
        makeGrammarDict: function(subject, object) {
          function posessive(name) {
            return name + "'s";
          }

          function fillDict(dict, target, prefix) {
            if (!prefix)
              prefix = "";

            if (target === viewer) {
              utils.extend(dict, utils.makePronounDict("you", prefix));
              dict[prefix + 'name'] = 'you';
              dict[prefix + 'name_pos'] = 'your';
            } else {
              utils.extend(dict, utils.makePronounDict(target.gender,
                                                       prefix));
              dict[prefix + 'name'] = target.definiteName;
              dict[prefix + 'name_pos'] = posessive(target.definiteName);
            }
          }

          var dict = {
            __subject: subject,
            __object: object
          };

          fillDict(dict, subject);
          fillDict(dict, object, "o_");
          return dict;
        },
        parseString: function(string, dict) {
          var isSubjPlural = (dict.__subject === viewer);
          var isObjPlural = (dict.__object === viewer);
          
          string = utils.parseString(string, isSubjPlural, isObjPlural);
          return utils.applyDict(string, dict);
        }
      };

      return self;
    }
  };
})();

File: grammar.py
import re

class View:
    def __init__(self, viewer):
        self.viewer = viewer
        
    def make_grammar_dict(self, subject, object):
        d = {}

        d["__subject"] = subject
        d["__object"] = object
        
        if subject == self.viewer:
            d.update(make_pronoun_dict("you"))
            d["name"] = "you"
            d["name_pos"] = "your"
        else:
            d.update(make_pronoun_dict(subject.get_sex()))
            d["name"] = subject.get_def_name()
            d["name_pos"] = make_name_posessive(d["name"])
            
        if object == self.viewer:
            d.update(make_pronoun_dict("you", "o_"))
            d["o_name"] = "you"
            d["o_name_pos"] = "your"            
        else:
            d.update(make_pronoun_dict(object.get_sex(), "o_"))
            d["o_name"] = object.get_def_name()
            d["o_name_pos"] = make_name_posessive(d["o_name"])
        return d
    
    def parse_grammar_string(self, stringy, grammar_dict):
        if grammar_dict["__subject"] == self.viewer:
            is_plural_1 = 1
        else:
            is_plural_1 = 0

        if grammar_dict["__object"] == self.viewer:
            is_plural_2 = 1
        else:
            is_plural_2 = 0

        stringy = parse_grammar_string(stringy, is_plural_1, is_plural_2)
        return stringy % grammar_dict

def make_name_posessive(name):
    if name[-1] == 's':
        return name + "'"
    else:
        return name + "'s"

def make_pronoun_dict(gender, prefix=None):
    if prefix == None:
        prefix = ""
    dicty = {}
    if gender == "male":
        himself = "himself"
        his = "his"
        him = "him"
        he = "he"
    elif gender == "female":
        himself = "herself"
        his = "her"
        him = "her"
        he = "she"
    elif gender == "neuter":
        himself = "itself"
        his = "its"
        him = "it"
        he = "it"
    elif gender == "you":
        himself = "yourself"
        his = "your"
        him = "you"
        he = "you"
    else:
        raise Exception, "Not a valid gender type."

    dicty['%shimself' % prefix] = himself
    dicty['%shis' % prefix] = his
    dicty['%shim' % prefix] = him
    dicty['%she' % prefix] = he
    
    return dicty

def parse_grammar_string(stringy, is_plural_1=0, is_plural_2=0):
    def parse_helper(stringy, restr, is_singular):
        match_obj = re.match(restr, stringy)
        if match_obj == None: return None
        if is_singular:
            stringy = match_obj.expand("\g<begin>\g<singular>\g<end>")
        else:
            if match_obj.group('plural') == None:
                stringy = match_obj.expand("\g<begin>\g<end>")
            else:
                stringy = match_obj.expand("\g<begin>\g<plural>\g<end>")
        return stringy

    # process first grammars (with []'s)
    restr = r"(?P<begin>.*)(\[(?P<singular>.*?)(\|(?P<plural>.+))?\])(?P<end>.*)"
    while 1:
        new_stringy = parse_helper(stringy, restr, not is_plural_1)
        if new_stringy == None:
            break
        else:
            stringy = new_stringy

    # process second grammars (with <>'s)
    restr = r"(?P<begin>.*)(\<(?P<singular>.*?)(\|(?P<plural>.+))?\>)(?P<end>.*)"
    while 1:
        new_stringy = parse_helper(stringy, restr, not is_plural_2)
        if new_stringy == None:
            break
        else:
            stringy = new_stringy
    
    return stringy
